lib/rules/camelcase.js

```diff
@@ -125,6 +125,38 @@ module.exports = {
             return false;
         }
 
+        /**
+         * Checks whether the given node represents assignment target property in destructuring.
+         *
+         * For examples:
+         *    ({a: b.foo} = c);  // => true for `foo`
+         *    ({a: b[foo]} = c); // => true for `foo`
+         *    ([a.foo] = b);     // => true for `foo`
+         *    ([a[foo]] = b);    // => true for `foo`
+         *    ({...a.foo} = b);  // => true for `foo`
+         *    ({...a[foo]} = b); // => true for `foo`
+         * @param {ASTNode} node An Identifier node to check
+         * @returns {boolean} True if the node is an assignment target property in destructuring.
+         */
+        function isAssignmentTargetPropertyInDestructuring(node) {
+            if (
+                node.parent.type === "MemberExpression" &&
+                node.parent.property === node
+            ) {
+                const effectiveParent = node.parent.parent;
+
+                return (
+                    effectiveParent &&
+                    effectiveParent.type === "Property" &&
+                    effectiveParent.value === node.parent &&
+                    effectiveParent.parent.type === "ObjectPattern" ||
+                    effectiveParent.type === "ArrayPattern" ||
+                    effectiveParent.type === "RestElement"
```

Comment:
This is same as:

```js
(effectiveParent &&
effectiveParent.type === "Property" &&
effectiveParent.value === node.parent &&
effectiveParent.parent.type === "ObjectPattern") ||
effectiveParent.type === "ArrayPattern" ||
effectiveParent.type === "RestElement"
```

so it will still crash if `effectiveParent` doesn't exist, instead of returning `false`.

I think we can actually just remove `effectiveParent &&`, because `MemberExpression` must have its `parent` node.